File: 567-permutation-in-string/permutation-in-string.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        vector<int> mp(26, 0);
        for(char ch : s1)
            mp[ch - 'a']++;
        
        int n1 = s1.size(), n2 = s2.size();
        vector<int> window(26, 0);
        for(int i = 0; i < n2; i++){
            if(i < n1)
                window[s2[i] - 'a']++;
            else{
                window[s2[i] - 'a']++;
                window[s2[i-n1] - 'a']--;
            }

            if(mp == window)
                return true;
        }
        return false;
    }
};

File: 567-permutation-in-string/permutation-in-string.java
class Solution {
    public boolean checkInclusion(String s1, String s2) {
        int[] freq = new int[26];
        for(char ch : s1.toCharArray())
            freq[ch - 'a']++;
        
        int l = 0;

        for(int h = 0; h < s2.length(); h++){
            freq[s2.charAt(h) - 'a']--;

            while(freq[s2.charAt(h) - 'a'] < 0){
                freq[s2.charAt(l) - 'a']++;
                l++;
            }
            if(h - l + 1 == s1.length()) return true;
        }
        return false;
    }
}